alg/teca_model_segmentation.py

```diff
@@ -69,51 +76,90 @@ def set_pred_name(self, name):
         """
         self.pred_name = name
 
-    def set_torch_inference_fn(self, torch_fn):
+    def set_inference_function(self, fn):
         """
-        set the final torch inference function. ex. torch.sigmoid()
+        set the final inference function. ex. torch.sigmoid()
         """
-        self.torch_inference_fn = torch_fn
+        self.inference_function = fn
 
     def input_preprocess(self, input_data):
         return input_data
 
     def output_postprocess(self, output_data):
         return output_data
 
-    def __set_transform_fn(self, fn, *args):
+    def set_threads_per_core(self, threads_per_core):
         """
-        if the data need to be transformed in a way then a function
-        could be provided to be applied on the requested data before
-        running it to the model.
+        Set the number of threds per core. Typically 1 gives the best
+        performance for CPU bound applicaitons, however 2 can sometimes
+        be useful. More than 2 will likeley harm performance.
         """
-        if not hasattr(fn, '__call__'):
-            raise TypeError(
-                "ERROR: The provided data transform function"
-                "is not a function"
-                )
-
-        if not args:
-            raise ValueError(
-                "ERROR: The provided data transform function "
-                "must at least have 1 argument -- the data array object to "
-                "apply the transformation on."
-                )
+        if threads_per_core > 4:
```

Comment:
what is the reason of choosing 4 threads per core here instead of 2 (as represented in the error message)?

Reply:
My observation with the test data so far is that 1 thread per core does the best, but the test data is small so I was thinking that 4 would leave room for some experimentation with over subscription. at the same time I wanted to catch the users who didn't realize this is a multiplicative factor. In your performance experiments what did you find?

Reply:
I found that setting more than 1 thread to torch speeds the process significantly... but the difference in performance of having 2, 4 & 8 threads isn't that big.

I ran on a month worth of data (month of april-2015), 30 files, 8 timesteps per file,  240 timesteps total, On Haswell:

| # of Nodes | # of ranks | # of torch threads | processing time (seconds) |
| ------------- | ------------- | ------------- | ------------- |
| 8  | 240  | 1  | 177 |
| 15  | 240  | 2  | 42 |
| 30  | 240  | 4  | 35 |
| 60  | 240  | 8  | 30 |

Reply:
you have a physical core available for each thread, this data does not describe performance for over subscription, which is what the setting is related to. if you don';t have that data , not a problem.

Reply:
Yeah makes sense.